GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,368 @@
+        }
+
+        public static IProductUpgrader Create(
+            ITracer tracer,
+            out string error)
+        {
+            NugetUpgraderConfig upgraderConfig = new NugetUpgraderConfig(tracer, new LocalGVFSConfig());
+            bool isConfigured;
+            bool isEnabled;
+
+            if (!upgraderConfig.TryLoad(out isEnabled, out isConfigured, out error))
+            {
+                return null;
+            }
+
+            if (!TryGetPersonalAccessToken(
+                GitBinPath,
+                upgraderConfig.FeedUrlForCredentials,
+                tracer,
+                out string token,
+                out error))
+            {
+                return null;
+            }
+
+            NuGetUpgrader upgrader = new NuGetUpgrader(
+                ProcessHelper.GetCurrentProcessVersion(),
+                tracer,
+                upgraderConfig,
+                ProductUpgraderInfo.GetAssetDownloadsPath(),
+                token);
+
+            return upgrader;
+        }
+
+        public bool UpgradeAllowed(out string message)
+        {
+            if (string.IsNullOrEmpty(this.nugetUpgraderConfig.FeedUrl))
+            {
+                message = "Nuget Feed URL has not been configured";
+                return false;
+            }
+
+            if (string.IsNullOrEmpty(this.nugetUpgraderConfig.PackageFeedName))
+            {
+                message = "URL to lookup credentials has not been configured";
+                return false;
+            }
+
+            if (string.IsNullOrEmpty(this.nugetUpgraderConfig.FeedUrlForCredentials))
+            {
+                message = "URL to lookup credentials has not been configured";
+                return false;
+            }
+
+            message = null;
+            return true;
+        }
+
+        public bool TryQueryNewestVersion(out Version newVersion, out string message)
+        {
+            newVersion = null;
+            message = null;
+
+            try
+            {
+                IList<IPackageSearchMetadata> queryResults = this.nugetFeedWrapper.QueryFeed(this.nugetUpgraderConfig.PackageFeedName).GetAwaiter().GetResult();
+
+                // Find the latest package
+                IPackageSearchMetadata highestVersion = null;
+                foreach (IPackageSearchMetadata result in queryResults)
+                {
+                    if (highestVersion == null || result.Identity.Version > highestVersion.Identity.Version)
+                    {
+                        highestVersion = result;
+                    }
+                }
+
+                if (highestVersion != null &&
+                    highestVersion.Identity.Version.Version > this.installedVersion)
+                {
+                    this.latestVersion = highestVersion;
+                }
+
+                newVersion = this.latestVersion?.Identity?.Version?.Version;
+
+                if (!(newVersion is null))

```

Comment:
`if (newVersion != null)`

Reply:
what is your reasoning here?

Reply:
Out of curiosity is there a difference between `== null` and `is null`? 

I have not seen `is` used with `null` before.

Reply:
> Out of curiosity is there a difference between == null and is null?

They can have different behavior depending on whether the equality operator is overloaded or not. The `is null` syntax makes it explicit what the code is doing. When using the equality operator, then it depends on what the equality operator does. There could also be a slight performance advantage to using the `is null` syntax when the equality operator is overloaded - it would depend on the implementation of the equality operator for that type.

Overall - I wouldn't expect an actual / noticeable difference here, but I think `is null` is the more technically appropriate syntax (at the cost of a bit more awkward syntax).

Reply:
My preference is for the syntax that @derrickstolee suggested `if (newVersion != null)`.

That's the syntax we're using elsewhere in the codebase and I personally find it easier to follow.